feat: add query-session-end and improve session-end events on Windows (#44598)
electron/electron: docs/api/base-window.md

Emitted when the window is closed. After you have received this event you should
remove the reference to the window and avoid using it any more.

#### Event: 'query-session-end' _Windows_

Returns:

* `event` [WindowSessionEndEvent][window-session-end-event]

Emitted when a session is about to end due to a shutdown, machine restart, or user log-off.
Calling `event.preventDefault()` can delay the system shutdown, though it’s generally best
to respect the user’s choice to end the session. However, you may choose to use it if
ending the session puts the user at risk of losing data.

#### Event: 'session-end' _Windows_

Returns:

* `event` [WindowSessionEndEvent][window-session-end-event]

> Emitted when system is in the process of shutting down after successfully handling 'query-session-end' event.

Reviewer: Wording suggestion. "after successfully handling" doesn't seem quite right, because the user can still force a shutdown even if you `event.preventDefault()` the `query-session-end` event.

```suggestion
Emitted when a session is about to end due to a shutdown, machine restart, or user log-off. Once this event fires, there is no way to prevent the session from ending.
```

Emitted when the window is closed. After you have received this event you should
remove the reference to the window and avoid using it any more.

#### Event: 'query-session-end' _Windows_

Returns:

* `event` Event
* `reasons` string[]

> Emitted when window session is going to end due to force shutdown or machine

Reviewer: This mentions "force shutdown". Does that mean `query-session-end` will fire again if the user elects to force a shutdown in the "These apps are preventing shutdown" screen? I would have intuitively expected that to trigger the `session-end` event instead.

Author: No, it won't I will remove "force" word.

Emitted when the window is closed. After you have received this event you should
remove the reference to the window and avoid using it any more.

#### Event: 'query-session-end' _Windows_

Returns:

* `event` Event

> * `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.

Reviewer 2: Our documentation parser will use string literals with the "Can be ..." syntax.
```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

Emitted when a session is about to end due to a shutdown, machine restart, or user log-off.
Calling `event.preventDefault(`) can delay the system shutdown, though it’s generally best
to respect the user’s choice to end the session. However, you may choose to use it if
ending the session puts the user at risk of losing data.

Unfortunately, Windows does not offer a way to differentiate between a shutdown and a reboot, meaning the 'shutdown'
reason is triggered in both scenarios. For more details on the `WM_QUERYENDSESSION` message and its associated reasons,
refer to the [MSDN documentation](https://learn.microsoft.com/en-us/windows/win32/shutdown/wm-queryendsession).

#### Event: 'session-end' _Windows_

Returns:

* `event` Event

> * `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.

Reviewer 2: ```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

Emitted when the window is closed. After you have received this event you should
remove the reference to the window and avoid using it any more.

> #### Event: 'session-end' _Windows_

Reviewer 2: One thing I missed is that we probably don't want to outright remove this event if it can be avoided.

@savely-krasovsky were you able to give @itsananderson's suggestion a try?
https://github.com/electron/electron/pull/44598#discussion_r1842998014

Author: @samuelmaddock could you approve my PR if everything is ok now?
